foreshadow/core/preparerstep.py

```diff
@@ -1,61 +1,157 @@
 """General base classes used across Foreshadow."""
+from collections import MutableMapping, defaultdict, namedtuple
+from inspect import signature
+
 from sklearn.base import BaseEstimator, TransformerMixin
 from sklearn.pipeline import Pipeline
 
 from foreshadow.core import logging
 from foreshadow.transformers.core import ParallelProcessor
 from foreshadow.transformers.core.notransform import NoTransform
-from foreshadow.transformers.core.pipeline import (
-    SingleInputPipeline,
-    TransformersPipeline,
+from foreshadow.transformers.core.pipeline import DynamicPipeline
+
+
+GroupProcess = namedtuple(
+    "GroupProcess", ["single_input", "step_inputs", "steps"]
 )
 
 
-def _check_parallelizable_batch(
-    column_mapping, group_number, PipelineClass=TransformersPipeline
-):
+class PreparerMapping(MutableMapping):
```

Comment:
Does it need to be a mutablemapping (I don't know), is there a version that allows the definition of a static version so that it doesn't get accidentally changed.

Reply:
I would see those steps getting modified by a generator passed into the init rather through the use of add in that case.

Reply:
Python doesn't have a builtin FrozenDict class, though I could always make one (it was proposed and rejected in PEP)

Reply:
#116  issue